applications/FluidDynamicsApplication/custom_processes/two_fluids_inlet_process.cpp

```diff
@@ -0,0 +1,243 @@
+    : Process(), mrInletModelPart(rModelPart) {
+
+    KRATOS_CHECK_VARIABLE_KEY( DISTANCE )
+    KRATOS_CHECK_VARIABLE_KEY( AUX_DISTANCE )
+    KRATOS_CHECK_VARIABLE_KEY( NODAL_AREA )
+
+    // checking all parameters that define the interface
+    Parameters default_parameters( R"(
+    {
+        "interface_normal"          : [0.0,1.0,0.0],
+        "point_on_interface"        : [0.0,0.25,0.0],
+        "inlet_transition_radius"   : 0.05
+    }  )" );
+
+    rParameters.ValidateAndAssignDefaults(default_parameters);
+
+    // finding the complete model of the inlet model part
+    ModelPart& rRootModelPart = mrInletModelPart.GetRootModelPart();
+
+    // setting the parameters to the private data members of the class
+    mInterfaceNormal = rParameters["interface_normal"].GetVector();
+    mInterfacePoint = rParameters["point_on_interface"].GetVector();
+    mInletRadius = rParameters["inlet_transition_radius"].GetDouble();
+
+    // normalization of itnerface normal vector
+    if ( norm_2( mInterfaceNormal ) > 1.0e-7 ){
+        mInterfaceNormal = mInterfacePoint / norm_2( mInterfaceNormal );
+    } else {
+        KRATOS_ERROR << "Error thrown in TwoFluidsInletProcess: 'interface_normal' in 'interface_settings' must not have a norm of 0.0." << std::endl;
+    }
+    // checking to avoid 0 transition radius (necessary for division)
+    if ( mInletRadius < 1.0e-7 ){
+        KRATOS_ERROR << "Error thrown in TwoFluidsInletProcess: 'inlet_transition_radius' in 'interface_settings' must not have a value smaller or equal 0.0." << std::endl;
+    }
+
+    // setting flags for the inlet on nodes and conditions
+    #pragma omp parallel for
+    for (int i_node = 0; i_node < static_cast<int>( mrInletModelPart.NumberOfNodes() ); ++i_node){
+        // iteration over all nodes
+        auto it_nodes = mrInletModelPart.NodesBegin() + i_node;
+        it_nodes->Set( INLET, true );
+    }
+    #pragma omp parallel for
+    for (int i_cond = 0; i_cond < static_cast<int>( mrInletModelPart.NumberOfConditions() ); ++i_cond){
+        // iteration over all conditions
+        auto it_cond = mrInletModelPart.ConditionsBegin() + i_cond;
+        it_cond->Set( INLET, true );
+    }
+
+    // Comment: The historical DISTANCE variable is used to compute a distance field that is then stored in a non-historical variable AUX_DISTANCE
+    //          The functions for the distance computation do not work on non-historical variables - this is reason for the following procedure (*)
+
+    // (*) temporally storing the distance field as an older version of itself (it can be assured that nothing is over-written at the start)
+    #pragma omp parallel for
+    for (int i_node = 0; i_node < static_cast<int>( rRootModelPart.NumberOfNodes() ); ++i_node){
+        // iteration over all nodes
+        auto it_node = rRootModelPart.NodesBegin() + i_node;
+        it_node->GetSolutionStepValue(DISTANCE, 2) = it_node->GetSolutionStepValue(DISTANCE, 0);
```

Comment:
In here you are assuming that you will always have 3 positions in the buffer (0, 1 and 2). Even though this is true for the current implementation of the two fluids element, which hardcodes the BDF2 scheme, it might not be the case in the future or the case in alternative formulations that might benefit from this process.

Since this is a potential and hard to debug source of problems, I would do something in here. My proposals are:

- We put a `KRATOS_ERROR_IF`macro that always checks that the buffer size is at least 3.
- We check the buffer size in the root model part and we temporally save the distance in its last position.

Reply:
Ok, I think, I see the point. 

const int buffer_size = r_root_model_part.GetBufferSize()
KRATOS_ERROR_IF( buffer_size < 2 ) << "TwoFluidsInletProcess: There is no space for an intermediate storage" << std::endl;
...
it_node->GetSolutionStepValue(DISTANCE, (buffer_size-1) ) = it_node->GetSolutionStepValue(DISTANCE, 0);